storm-core/test/jvm/org/apache/storm/daemon/supervisor/BasicContainerTest.java

```diff
@@ -0,0 +1,459 @@
+    public void testRunProfiling() throws Exception {
+        final long pid = 100;
+        final String topoId = "test_topology";
+        final int port = 8080;
+        final String workerId = "worker-id";
+        final String stormLocal = ContainerTest.asAbsPath("tmp", "testing");
+        final String topoRoot = ContainerTest.asAbsPath(stormLocal, topoId, String.valueOf(port));
+        final File workerArtifactsPid = ContainerTest.asAbsFile(topoRoot, "worker.pid");
+        
+        final Map<String, Object> superConf = new HashMap<>();
+        superConf.put(Config.STORM_LOCAL_DIR, stormLocal);
+        superConf.put(Config.STORM_WORKERS_ARTIFACTS_DIR, stormLocal);
+        
+        LocalAssignment la = new LocalAssignment();
+        la.set_topology_id(topoId);
+        
+        AdvancedFSOps ops = mock(AdvancedFSOps.class);
+        when(ops.slurpString(workerArtifactsPid)).thenReturn(String.valueOf(pid));
+        
+        LocalState ls = mock(LocalState.class);
+        
+        MockBasicContainer mc = new MockBasicContainer(ops, port, la, superConf, 
+                new HashMap<String, Object>(), "SUPERVISOR", null, ls, "profile");
+        mc._workerId = workerId;
+        
+        //HEAP DUMP
+        ProfileRequest req = new ProfileRequest();
+        req.set_action(ProfileAction.JMAP_DUMP);
+        
+        mc.runProfiling(req, false);
+        
+        assertEquals(1, mc.profileCmds.size());
+        CommandRun cmd = mc.profileCmds.get(0);
+        mc.profileCmds.clear();
+        assertEquals(Arrays.asList("profile", String.valueOf(pid), "jmap", topoRoot), cmd.cmd);
+        assertEquals(new File(topoRoot), cmd.pwd);
+        
+        //JSTACK DUMP
+        req.set_action(ProfileAction.JSTACK_DUMP);
+        
+        mc.runProfiling(req, false);
+        
+        assertEquals(1, mc.profileCmds.size());
+        cmd = mc.profileCmds.get(0);
+        mc.profileCmds.clear();
+        assertEquals(Arrays.asList("profile", String.valueOf(pid), "jstack", topoRoot), cmd.cmd);
+        assertEquals(new File(topoRoot), cmd.pwd);
+        
+        //RESTART
+        req.set_action(ProfileAction.JVM_RESTART);
+        
+        mc.runProfiling(req, false);
+        
+        assertEquals(1, mc.profileCmds.size());
+        cmd = mc.profileCmds.get(0);
+        mc.profileCmds.clear();
+        assertEquals(Arrays.asList("profile", String.valueOf(pid), "kill"), cmd.cmd);
+        assertEquals(new File(topoRoot), cmd.pwd);
+        
+        //JPROFILE DUMP
+        req.set_action(ProfileAction.JPROFILE_DUMP);
+        
+        mc.runProfiling(req, false);
+        
+        assertEquals(1, mc.profileCmds.size());
+        cmd = mc.profileCmds.get(0);
+        mc.profileCmds.clear();
+        assertEquals(Arrays.asList("profile", String.valueOf(pid), "dump", topoRoot), cmd.cmd);
+        assertEquals(new File(topoRoot), cmd.pwd);
+        
+        //JPROFILE START
+        req.set_action(ProfileAction.JPROFILE_STOP);
```

Comment:
Ya it was really confusing until I asked @kishorvpatil about it and he explained it all to me.